lookout/style/typos/analyzer.py

```diff
@@ -145,17 +155,69 @@ def render_comment_text(self, typo_fix: TypoFix) -> str:
+        for final_sugg, conf in self.generate_identifier_suggestions(
+                typo_fix.candidates, typo_fix.token, n_candidates=self.config["n_candidates"],
+        ):
+            suggestions.append(self.reconstruct_identifier(self.parser, final_sugg,
+                                                           typo_fix.token))
+            confidences.append(conf)
+        return self.comment_template.render(
+            identifier=typo_fix.token,
+            suggestions=suggestions,
+            old_code_line=file_lines[typo_fix.line_number - 1],
+            confidences=self._normalize_confidences(confidences),
+            zip=zip,
+            enumerate=enumerate,
+            int=int,
+            )
+
+    @staticmethod
+    def _compute_confidence(confidences: Sequence[float]) -> float:
+        confidence = 1
+        for conf in confidences:
+            confidence *= conf
+        return confidence
 
     @staticmethod
-    def _get_comment_confidence(suggestions: Sequence[Candidate]) -> int:
-        invert_confidence = 1
-        for suggestion in suggestions:
-            invert_confidence *= (1 - suggestion.confidence)
-        return int(100 * (1 - invert_confidence))
+    def _normalize_confidences(confidences: Sequence[float]):
+        s = sum(confidences)
+        return [conf / s for conf in confidences]
+
+    def generate_identifier_suggestions(self, suggestions: Mapping[str, Iterable[Candidate]],
+                                        identifier: str, n_candidates: int = 3) \
+            -> Iterator[Tuple[List[str], float]]:
+        """
+        Generate suggestions for identifier and compute probability of suggestion.
+
+        :param suggestions: suggestions - mapping from token to list of candidates.
+        :param identifier: initial identifier.
+        :param n_candidates: number of candidates to generate.
+        :return: generator of (list of suggestion for identifier, probability).
+        """
+        initial_tokens = self.parser.split(identifier)
+        final_sugg = []
+        confidences = [1]
+        suggestion_queue = []
+        for i, t in enumerate(initial_tokens):
+            try:
+                final_sugg.append(suggestions[t][0].token)
+                confidences.append(suggestions[t][0].confidence)
+                for sugg in suggestions[t][1:]:
+                    suggestion_queue.append((i, sugg.token, sugg.confidence))
+            except KeyError:
+                final_sugg.append(t)
+        yield final_sugg, self._compute_confidence(confidences)
+
+        suggestion_queue = list(reversed(sorted(suggestion_queue, key=itemgetter(2))))
+        for j in range(n_candidates - 1):
+            pos = suggestion_queue[j][0]
+            sugg = suggestion_queue[j][1]
+            conf = suggestion_queue[j][2]
+            final_sugg[pos] = sugg
+            confidences[pos] = conf
+            yield final_sugg, self._compute_confidence(confidences)
```

Comment:
I spent quite a lot of time trying to understand the logic of the last (203-220) lines.
I believe there should be a more straightforward way to do the same. I think that `suggestions` have tokens for the provided identifier only, so we can use `product` from itertools. https://docs.python.org/3/library/itertools.html#itertools.product to do the same. 

So I'd suggest you iterate through the loop of the product of changes suggestions.  

And a question. Does suggestion list contain an original token with a probability (so a suggestion to not change this token)?

Reply:
`itertools.product` will have high computational complexity - `n_tokens ^ n_candidates` + sorting them to select best candidates so using big-O notation will give use:
* `O(product + sorting)` -> `O(n_tokens ^ n_candidates + n_tokens ^ n_candidates * log(n_tokens ^ n_candidates))`.

The current solution has 
* `O(select_best_candidates + sort_rest_candidates)` -> `O(n_tokens + n_tokens * n_candidates * log(n_tokens * n_candidates))` 

So the current solution is much better in terms of computational complexity.

Reply:
ok, can you add some comments, because now I am not sure I understand the logic?

Reply:
About logic - read it carefully - if you still have a question - call me.
About question - @irinakhismatullina knows the answer.

Reply:
Should everyone call you who do not understand the logic? ;)
It is good to have comments for every unobvious code. Because you increase code reading speed and it is imprortant.

Reply:
The original token is not present yet. Before it would be like *why the hell are you suggesting me to change this tokens to itself*. So i think we should add it, but as a separate info, not like one of the candidates. I'll think of the best way to do it. How urgent is it?

Reply:
\+1 for adding a comment about the logic in any case

Reply:
Thanks, @irinakhismatullina!
What we need is an ability to get a probability that token is correct and should not be fixed. Otherwise, we can have such a situation: 

Identifier: TupoThen
Suggestions: 
1. Tupo -> Typo (with prob 0.9)
2. Then -> Than (with prop 0.1)

So we do not have an option to change only `Tupo` and keep `Then`.

The initial token itself we can get from https://github.com/src-d/style-analyzer/pull/709/files#diff-18972d7c14aaaa30c48c1560b2e149c0R204

Reply:
In the example you provided most likely the corrector would not suggest to correct `Then` at all (it makes only the suggestions that have higher proba, than the proba of the token itself). Anyway this must be added, but used carefully.

Reply:
\-1 for being noisy in our predictions

Reply:
https://github.com/src-d/style-analyzer/blob/c93cb3147d604fa1be7da7a763963cb5e78b56f3/lookout/style/typos/analyzer.py#L190~L223

Reply:
> it makes only the suggestions that have higher proba, than the proba of the token itself

ok, good!